### PR TITLE
feat: add follow and unfollow user implementation on highlights

### LINKED ISSUE #885: feat: following user on highlights
220161799-951dae81-b9a8-4e17-af0b-834a2eab548f (1)

related: https://github.com/open-sauced/insights/issues/715

### LINKED ISSUE #1076: Feature: unfollow should be "following"
### Type of feature

🍕 Feature

### Current behavior

The following state of the button is a CTA to unfollow

Screen Shot 2023-04-03 at 5 26 29 PM

### Suggested solution

Change the copy to "Following"

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [ ] I agree to follow this project's Contribution Docs

### PR DESCRIPTION
**Pull request by OgDev-01**
## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Description
This PR: 

- Add the follow and unfollow functionality to user highlights
- redirects unauthenticated users to authenticate when they try to follow

## Related Tickets & Documents
Closes #885 
Closes #1076 

### Logged in owner
image

### user not followed
image

### followed user
image

image

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [x] 🙅 no documentation needed

### CODE REVIEW

File: components/molecules/ContributorHighlight/contributor-highlight-card.tsx [hunk 1]
```diff
@@ -218,12 +221,28 @@ const ContributorHighlightCard = ({
...
-                  </div>
-                </DropdownMenuItem>
+                {loggedInUser ? (
+                  <DropdownMenuItem
+                    className={`rounded-md ${loggedInUser.user_metadata.user_name === user && "hidden"}`}
+                  >
+                    <div onClick={isError ? follow : unFollow} className="flex gap-2.5 py-1  items-center pl-3 pr-7">
+                      <FaUserPlus size={22} />
+                      <span>
+                        {!isError ? "unfollow" : "follow"} {user}
```
Comment (brandonroberts) [inline components/molecules/ContributorHighlight/contributor-highlight-card.tsx:231]: Capitalize actions. Do we need the username after?

```suggestion
                        {!isError ? "Unfollow" : "Follow"} {user}
```

**Review by brandonroberts — CHANGES_REQUESTED** [review]

File: components/molecules/ContributorHighlight/contributor-highlight-card.tsx [hunk 2]
```diff
@@ -218,12 +221,28 @@ const ContributorHighlightCard = ({
...
+                    </div>
+                  </DropdownMenuItem>
+                ) : (
+                  <DropdownMenuItem className="rounded-md">
+                    <div
+                      onClick={async () => await signIn({ provider: "github" })}
+                      className="flex gap-2.5 py-1  items-center pl-3 pr-7"
+                    >
+                      <FaUserPlus size={22} />
+                      <span>Follow {user}</span>
```
[the remark this answers is not in the document]
Comment (OgDev-01) [reply components/molecules/ContributorHighlight/contributor-highlight-card.tsx:242]: Don't get the question here?

[the remark this answers is not in the document]
Comment (brandonroberts) [reply components/molecules/ContributorHighlight/contributor-highlight-card.tsx:242]: Do we need the ` {user}`  after the "Follow" text

[the remark this answers is not in the document]
Comment (OgDev-01) [reply components/molecules/ContributorHighlight/contributor-highlight-card.tsx:242]: I'm not sure which is better, but leaving the gives information about who you want to follow
And that's how it was plotted in the design on figma

File: components/molecules/ContributorHighlight/contributor-highlight-card.tsx (hunk 1 above)
Comment (OgDev-01) [reply components/molecules/ContributorHighlight/contributor-highlight-card.tsx:231]: Yes we do as seen here
image

File: components/molecules/ContributorHighlight/contributor-highlight-card.tsx [hunk 3]
```diff
@@ -70,12 +71,14 @@ const ContributorHighlightCard = ({
...
   const [errorMsg, setError] = useState("");
   const [loading, setLoading] = useState(false);
-  const { user: loggedInUser } = useSupabaseAuth();
+  const { user: loggedInUser, sessionToken, signIn } = useSupabaseAuth();
   const [openEdit, setOpenEdit] = useState(false);
   const [alertOpen, setAlertOpen] = useState(false);
   const [deleteLoading, setDeleteLoading] = useState(false);
   const [host, setHost] = useState("");
 
+  const { follow, unFollow, isError } = useFollowUser(user, sessionToken || "");
```
[the remark this answers is not in the document]
Comment (OgDev-01) [reply components/molecules/ContributorHighlight/contributor-highlight-card.tsx:80]: Thanks for this. Updating.... 🏃

**Review by brandonroberts — APPROVED** [review]

File: components/molecules/ContributorHighlight/contributor-highlight-card.tsx [hunk 4]
```diff
@@ -218,12 +221,28 @@ const ContributorHighlightCard = ({
...
-                  </div>
-                </DropdownMenuItem>
+                {loggedInUser ? (
+                  <DropdownMenuItem
+                    className={`rounded-md ${loggedInUser.user_metadata.user_name === user && "hidden"}`}
+                  >
+                    <div onClick={isError ? follow : unFollow} className="flex gap-2.5 py-1  items-center pl-3 pr-7">
+                      <FaUserPlus size={22} />
+                      <span>
+                        {!isError ? "Unfollow" : "Follow"} {user}
```
Comment (Deadreyo) [inline components/molecules/ContributorHighlight/contributor-highlight-card.tsx:231]: So basically `isError` actually stands for whether the person is followed or not? If that's the case, it needs a clearer name from the hook itself to be able to quickly understand this.

**Review by Deadreyo — CHANGES_REQUESTED** [review]

File: components/molecules/ContributorHighlight/contributor-highlight-card.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (bdougie) [reply components/molecules/ContributorHighlight/contributor-highlight-card.tsx:242]: I am for the user name. Give folks context even though they are looking at the highlight of the user. Better to have them not guess.

**Review by bdougie — APPROVED** [review]

**Pull request merged**